File: apprefurbished/src/main/java/com/androidcollider/easyfin/home/HomeModel.java
package com.androidcollider.easyfin.home;

import androidx.core.util.Pair;

import com.androidcollider.easyfin.common.repository.Repository;

import java.util.Map;

import io.reactivex.rxjava3.core.Flowable;

/**
 * @author Ihor Bilous
 */

class HomeModel implements HomeMVP.Model {

    private Repository repository;

    HomeModel(Repository repository) {
        this.repository = repository;
    }

    @Override
    public Flowable<Pair<Map<String, double[]>, Map<String, double[]>>> getBalanceAndStatistic(int statisticPosition) {
        return Flowable.combineLatest(
                repository.getAccountsAmountSumGroupByTypeAndCurrency(),
                repository.getTransactionsStatistic(statisticPosition),
                Pair::new);
    }

    @Override
    public Flowable<Map<String, double[]>> getBalance() {
        return repository.getAccountsAmountSumGroupByTypeAndCurrency();
    }

    @Override
    public Flowable<Map<String, double[]>> getStatistic(int statisticPosition) {
        return repository.getTransactionsStatistic(statisticPosition);
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/home/HomeModel.kt
package com.androidcollider.easyfin.home

import com.androidcollider.easyfin.common.repository.Repository
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
internal class HomeModel(private val repository: Repository) : HomeMVP.Model {

    override fun getBalanceAndStatistic(statisticPosition: Int):
            Single<Pair<Map<String, DoubleArray>, Map<String, DoubleArray>>> {
        return Single.zip(
            repository.accountsAmountSumGroupByTypeAndCurrency!!,
            repository.getTransactionsStatistic(statisticPosition)!!
        ) { first: Map<String, DoubleArray>, second: Map<String, DoubleArray> ->
            Pair(first, second)
        }
    }

    override val balance: Single<Map<String, DoubleArray>>
        get() = repository.accountsAmountSumGroupByTypeAndCurrency!!

    override fun getStatistic(statisticPosition: Int): Single<Map<String, DoubleArray>> {
        return repository.getTransactionsStatistic(statisticPosition)!!
    }
}